controllers/metal3.io/host_state_machine.go

```diff
@@ -443,6 +447,28 @@ func (hsm *hostStateMachine) provisioningCancelled() bool {
 	return false
 }
 
+func (hsm *hostStateMachine) customDeployProvisioningCancelled() bool {
+	if hsm.Host.Spec.CustomDeploy == nil {
+		return true
+	}
+	if hsm.Host.Spec.CustomDeploy.Method == "" {
+		return true
+	}
+	if hsm.Host.Status.Provisioning.CustomDeploy == nil {
+		return false
+	}
+	if hsm.Host.Status.Provisioning.CustomDeploy.Method == "" {
+		return false
+	}
+	if hsm.Host.Spec.CustomDeploy.Method != hsm.Host.Status.Provisioning.CustomDeploy.Method {
+		return true
+	}
+	if hsm.Host.Spec.Image != nil && hsm.Host.Spec.Image.URL != hsm.Host.Status.Provisioning.Image.URL {
```

Comment:
Looks like this one should be already covered by `imageProvisioningCancelled()`?

Reply:
It's not. The two functions are called with `&&` between them, so if `imageProvisioningCancelled` returns `true`, this will return `false` and the result will be `false`.

Reply:
Yes, that's right, but it looks a little bit tricky, and since this method could trigger a disruptive  action like deprovisioning I think it's worth trying to simplify it as much as possible.
I guess that the goal of the `provisioningCancelled()` method is essentially to detect a change - during the `Provisioning` or `Provisioned` state - such that invalidates the current state, thus requiring a deprovisioning.

As per the proposal, the  `Image` field should be considered optional if `CustomDeploy` is set (but not if it isn't of course), and this may suggest to:
1)  Check first `CustomDeploy` and then `Image` 
    * Do not consider `Image` nullity as a trigger condition in such case if `CustomDeploy` is set
2) If `CustomDeploy` is nil, check `Image` field as before to ensure the previous behavior

Ie, something like that: 

```
func (hsm *hostStateMachine) provisioningCancelled() bool {

	if hsm.Host.Spec.CustomDeploy != nil {
		if hsm.Host.Spec.CustomDeploy.Method == "" {
			return true
		}
		if hsm.Host.Status.Provisioning.CustomDeploy != nil && hsm.Host.Spec.CustomDeploy.Method != hsm.Host.Status.Provisioning.CustomDeploy.Method {
			return true
		}
		// At this point spec.CustomDeploy value didn't change (and it's not empty). Only a discrepancy in the Image
		// could require a deprovisioning, but if it isn't set or empty then is fine
		if hsm.Host.Spec.Image != nil && hsm.Host.Status.Provisioning.Image.URL != "" && hsm.Host.Spec.Image.URL != hsm.Host.Status.Provisioning.Image.URL {
			return true
		}

		return false
	}

	// Fallback to the previous case
	return hsm.imageProvisioningCancelled()
}
```

Reply:
I also read this wrong, and I have created truly terrible bugs in the past (see #915) by getting this function wrong, so I agree that we should seek to simplify the logic.

Reply:
Will do, but the proposed snippet doesn't cover one case (removing custom deploy while keeping the image). I'll try to figure it out.